AutoMerge/src/util.jl

```diff
@@ -75,6 +77,192 @@ function parse_registry_pkg_info(registry_path, pkg, version=nothing)
     return (; uuid=uuid, repo=repo, subdir=subdir, tree_hash=tree_hash)
 end
 
+#####
+##### Version diff functionality
+#####
+
+"""
+    find_previous_semver_version(pkg::String, current_version::VersionNumber, registry_path::String) -> Union{VersionNumber, Nothing}
+
+Finds the previous semver version for a package. Returns the maximum version that is less than the current version,
+or `nothing` if there are no previous versions.
+"""
+function find_previous_semver_version(pkg::String, current_version::VersionNumber, registry_path::String)
+    all_pkg_versions = all_versions(pkg, registry_path)
+    previous_versions = filter(v -> v < current_version, all_pkg_versions)
+    return isempty(previous_versions) ? nothing : maximum(previous_versions)
+end
+
+"""
+    tree_sha_to_commit_sha(tree_sha::String, clone_dir::String; subdir::String="") -> Union{String, Nothing}
+
+Converts a git tree SHA to a commit SHA by finding a commit that has that tree.
+Returns the commit SHA string, or `nothing` if no commit is found.
+"""
+function tree_sha_to_commit_sha(tree_sha::AbstractString, clone_dir::AbstractString; subdir::AbstractString = "")
+    # Normalize to a full tree object ID; return nothing if it’s not a tree reachable in this repo
+    full_tree = try
+        # --verify fails if not found; --quiet suppresses stderr noise
+        sha_cmd = "$tree_sha^{tree}"
+        readchomp(Cmd(`git -c gc.auto=0 rev-parse --verify --quiet $sha_cmd`;
+                      dir=clone_dir, env=Dict("GIT_NO_REPLACE_OBJECTS"=>"1")))
```

Comment:
There won't  be any replace objects in our clone of the repository, so `env` doesn't do anything. `dir=clone_dir` can be replaced with `git -C $(clone_dir)`. Then the `Cmd` wrapper can be skipped.